diff --git a/sys/rpicamsrc/meson.build b/sys/rpicamsrc/meson.build
index dc18a8345..fcc18ef18 100644
--- a/sys/rpicamsrc/meson.build
+++ b/sys/rpicamsrc/meson.build
@@ -34,7 +34,7 @@ if not cc.has_header('bcm_host.h', args: rpi_inc_args)
 endif
 
 mmal_deps = []
-foreach rpi_lib : ['mmal_core', 'mmal_util', 'mmal_vc_client', 'vcos', 'bcm_host']
+foreach rpi_lib : ['mmal_core', 'mmal_util', 'mmal_vc_client', 'vcos', 'bcm_host', 'vchostif']
   l = cc.find_library(rpi_lib, dirs: rpi_lib_path, required: false)
   if not l.found()
     if get_option('rpicamsrc').enabled()
